### DISCUSSION TITLE
LeNet5 Example from Brevitas

### DISCUSSION THREAD

**Discussion by ghost** [Xilinx/finn#809] [Q&A] [answered]
Hi,

i went through all examples from the notebooks and wanted to deploy LeNet5 on the Pynq-Z1 from scratch. I took the LeNet5 example from the Getting-Started from Brevitas(https://xilinx.github.io/brevitas/getting_started.html), see the code below.

```python
from torch import nn, randn
from torch.nn import Module

import torch.nn.functional as F

import brevitas.nn as qnn
import brevitas.onnx as bo

class QuantWeightLeNet(Module):
    def __init__(self):
        super(QuantWeightLeNet, self).__init__()
        self.conv1 = qnn.QuantConv2d(3, 6, 5, bias=True, weight_bit_width=4)
        self.relu1 = nn.ReLU()
        self.conv2 = qnn.QuantConv2d(6, 16, 5, bias=True, weight_bit_width=4)
        self.relu2 = nn.ReLU()
        self.fc1   = qnn.QuantLinear(16*5*5, 120, bias=True, weight_bit_width=4)
        self.relu3 = nn.ReLU()
        self.fc2   = qnn.QuantLinear(120, 84, bias=True, weight_bit_width=4)
        self.relu4 = nn.ReLU()
        self.fc3   = qnn.QuantLinear(84, 10, bias=True, weight_bit_width=4)

    def forward(self, x):
        out = self.relu1(self.conv1(x))
        out = F.max_pool2d(out, 2)
        out = self.relu2(self.conv2(out))
        out = F.max_pool2d(out, 2)
        out = out.reshape(out.shape[0], -1)
        out = self.relu3(self.fc1(out))
        out = self.relu4(self.fc2(out))
        out = self.fc3(out)
        
        return out
    
def export(model, ishape, chkpt_name):
    BrevitasONNXManager.export(model, ishape, chkpt_name)
    qonnx_cleanup(chkpt_name, out_file=chkpt_name)
    model = ModelWrapper(chkpt_name)
    model = model.transform(ConvertQONNXtoFINN())
    model.save(chkpt_name)

quant_weight_lenet = QuantWeightLeNet()

bo.export_finn_onnx(quant_weight_lenet, (1, 3, 32, 32), "lenet5_quant_weights.onnx")
```

Creating and exporting worked without any problems. My starting point is the following onnx graph:

<img src="https://user-images.githubusercontent.com/59102836/235926464-3c6d86c5-23b4-4d47-b99a-5b50eaac7e8a.png" height="2000">

The first steps were to clean up the NN, divide the inputs by 255, add a topk node, and clean up the NN again. The resulting NN looks like this:

<img src="https://user-images.githubusercontent.com/59102836/235939518-8c4c875e-bfa4-4b4a-bc0a-2b00e12906b8.png" height="2000">

After that, I started the streamline transformation with the following transformation steps:

```python
    streamline_config_linear = [
        MoveScalarLinearPastInvariants,
        Streamline,
        LowerConvsToMatMul,
        MakeMaxPoolNHWC,
        AbsorbTransposeIntoMultiThreshold,
        Streamline,
        InferDataLayouts, 
        RemoveUnusedTensors,
    ]
```

The result of the streamlining looks like this:

<img src="https://user-images.githubusercontent.com/59102836/235939946-b03a6738-2030-43a2-a294-38e804785a29.png" height="2000">

I think I have a lot of nodes that can't be converted. But now I don't know what to do. Is it because of how the NN is constructed (only the weights are quantized)?  

What are the minimum requirements for the NN in terms of quantization and word width? 

Do i need a binary neural network?

Many thanks in advance,
Michael

**Comment by MFJI** [comment]
Hello Michael,

I think the 4-bit weight works for FINN. FINN can handle up to 8 bits from my experience. But I think the nn.ReLU doesn't work for FINN. I think maybe you need to use qnn.QuantReLU instead.

Good luck.
Mengfei

**Comment by auphelia** [comment, marked as answer]
Hi @devthrough ,

Like @MFJI already said, the quantization is not the issue here. Before we can convert a network into HLS layers and generate hardware we need to eliminate **all** the floating-point operations in the graph with streamlining. I see that you already started to look into this and applied some of the streamlining functions. Some networks need multiple rounds of these transformations to make sure that the fp operations are moved in front of MultiThreshold to then be able to merge them into the thresholds.

You can have a look at the [tfc or cnv end2end examples](https://github.com/Xilinx/finn/tree/dev/notebooks/end2end_example), as well as the build steps in finn-examples, e.g. the custom step for [MobileNet-v1](https://github.com/Xilinx/finn-examples/blob/main/build/mobilenet-v1/custom_steps.py#L53) to learn more about this.
